src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,123 @@
 package core.basesyntax;
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int DEFAULT_CAPACITY = 16;
+    private static final float LOAD_FACTOR = 0.75f;
+
+    private Node<K, V>[] table;
+    private int size;
+    private int capacity;
+    private int threshold;
+
+    public static class Node<K, V> {
+        private K key;
+        private V value;
+        private Node<K, V> next;
+
+        Node(K key, V value, Node<K, V> next) {
+            this.key = key;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    public MyHashMap() {
+        this.capacity = DEFAULT_CAPACITY;
+        this.table = (Node<K, V>[]) new Node[capacity];
+        this.size = 0;
+        this.threshold = (int) (capacity * LOAD_FACTOR);
+    }
+
+    private int hash(K key) {
+        if (key == null) {
+            return 0;
+        }
+        int hashCode = key.hashCode();
+        return hashCode == Integer.MIN_VALUE ? 0 : Math.abs(hashCode);
+    }
+
+    private void resize() {
+        capacity *= 2;
+        threshold = (int) (capacity * LOAD_FACTOR);
+        Node<K, V>[] newTable = (Node<K, V>[]) new Node[capacity];
+
+        for (Node<K, V> node : table) {
+            while (node != null) {
+                int newIndex = hash(node.key) % capacity;
+                Node<K, V> next = node.next;
+
+                node.next = newTable[newIndex];
+                newTable[newIndex] = node;
+
+                node = next;
+            }
+        }
+
+        table = newTable;
+    }
 
     @Override
     public void put(K key, V value) {
+        if (key == null) {
+            if (table[0] == null) {
+                table[0] = new Node<>(null, value, null);
+                size++;
+            } else {
+                table[0].value = value;
+            }
+            return;
+        }
+
+        int index = hash(key) % capacity;
+        Node<K, V> current = table[index];
 
+        if (current == null) {
+            table[index] = new Node<>(key, value, null);
+            size++;
+            if (size >= threshold) {
+                resize();
+            }
+            return;
+        }
+
+        while (true) {
+            if (current.key == null ? key == null : current.key.equals(key)) {
```

Comment:
Ensure that `current.key` is checked for null before calling `equals` to prevent a potential `NullPointerException`. Consider using a null-safe comparison like `Objects.equals(current.key, key)`.